serialization/cbor.ksy

```diff
@@ -0,0 +1,328 @@
+seq:
+  - id: ib
+    type: ib
+  - id: payload
+    type:
+      switch-on: ib.major_type
+      cases:
+        'type::uint': uint
+        'type::neg_uint': uint
+        'type::byte_str': byte_str
+        'type::text_str': text_str
+        'type::array': array
+        'type::map': map
+        'type::tag': tag
+        'type::simple_or_float': simple_or_float
+types:
+  ib:
+    seq:
+      - id: major_type
+        type: b3
+        enum: type
+      - id: additional_info
+        type: b5
+    instances:
+      is_raw_value:
+        value: additional_info < 24
+      value:
+        value: additional_info
+        if: is_raw_value
+      is_extension:
+        value: additional_info == 24
+      is_integer_length:
+         value: "additional_info >= 24 and additional_info <= 27"
+      integer_length:
+        value: additional_info - 24
+        if: is_integer_length
+      is_reserved:
+        value: "additional_info >= 28 and additional_info <= 30"
+      is_indefinite_length:
+        value: additional_info == 31
+  uint:
+    seq:
+      - id: value_with_length
+        type:
+          switch-on: _parent.as<cbor>.ib.integer_length
+          cases:
+            0: u1
+            1: u2
+            2: u4
+            3: u8
+        if: _parent.as<cbor>.ib.is_integer_length
+    instances:
+      is_negative:
+        value: "_parent.as<cbor>.ib.major_type == type::neg_uint"
+      raw_value:
+        value: "(_parent.as<cbor>.ib.is_raw_value?_parent.as<cbor>.ib.value:value_with_length)"
+      modulus:
+        value: "(is_negative?1+raw_value:raw_value)"
+      value:
+        value: "(is_negative?-modulus:modulus)"
+  byte_str:
+    seq:
+      - id: length
+        type: uint
+      - id: value
+        size: length.value
+    instances:
+      ib:
+        value: _parent.ib
+  text_str:
+    seq:
+      - id: length
+        type: uint
+      - id: value
+        type: str
+        encoding: utf-8
+        size: length.value
+        doc: "No escaping is used."
+    instances:
+      ib:
+        value: _parent.ib
+  array:
+    seq:
+      - id: length
+        type: uint
+      - id: items
+        type: cbor
+        repeat: expr
+        repeat-expr: length.value
+    instances:
+      ib:
+        value: _parent.ib
+  map:
+    seq:
+      - id: length
+        type: uint
+      - id: items
+        type: pair
+        repeat: expr
+        repeat-expr: length.value
+    instances:
+      ib:
+        value: _parent.ib
+    types:
+      pair:
+        seq:
+          - id: key
+            type: cbor
+          - id: value
+            type: cbor
+  simple_or_float:
+    seq:
+      - id: value
+        type:
+          switch-on: type
+          cases:
+            "type::simple": simple_or_float
+            #"type::f2": f2
+            "type::f4": f4
+            "type::f8": f8
+    instances:
+      type:
+        value: _parent.as<cbor>.ib.additional_info
```

Comment:
I keep saying the same: **don't use casting unless you're sure what you're doing**. In ideal case, you shouldn't need to use casting at all. There are only few exotic cases for valid use of typecasting, in which the compiler has no way of knowing that the attribute will have some specific type. _(One of such cases is for example https://github.com/kaitai-io/kaitai_struct/issues/200.)_

Otherwise the casting will just suppress the errors when you're trying to do something clearly unreachable, and it will give both the compiler and the developer a false sense of security that the expression will be always of some type. Therefore you should always think twice if you need to use casting and **why** is it necessary.

This is also the case. Apparently you added the cast `.as<cbor>`, because simple `_parent.ib.additional_info` doesn't work. That makes sense, because the `_parent` in type `simple_or_float` can be of both types `cbor` and `simple_or_float` (you are using the type recursively in the `value` field a few lines above). Next time, please think about why is it necessary to use casting and if it works in all cases.

If you really need use `simple_or_float` recursively, I suggest to make it parametric and pass the `type` attribute to it from the parent struct. You'll get rid of `_parent` usage and if the `type` is `"type::simple"`, you can tell the descendant which type it should use.